falcon/media/handlers.py

```diff
@@ -54,27 +109,27 @@ def _resolve_media_type(self, media_type, all_media_types):
 
         return resolved
 
+    # NOTE(kgriffs): Most apps will probably only use one or two media handlers,
+    #   but we use maxsize=64 to give us some wiggle room just in case someone
+    #   is using versioned media types or something, and to cover various
+    #   combinations of the method args. We may need to tune this later.
+    @functools.lru_cache(maxsize=64)
     def find_by_media_type(self, media_type, default, raise_not_found=True):
```

Comment:
As discussed on Gitter, we should IMHO simply define new methods (`_find_async_handler` or similar etc) that we need, and make them private. And leave this one as-is.

Otherwise we need to mention the updated return signature as a breaking change, since this was not an underscored method, and the class itself is public. (NB we would have not given any release inbetween to ease this breaking change either.)

Reply:
I actually agree, it's just something I forgot to do.

Reply:
To avoid repetition, I think this method could call the new one and just return the first argument, so we avoid having two implementations that do the same thing.

Reply:
That's a good point; doing so would work as the implementation of `resolve()` currently stands. I took a "safety first", non-DRY approach to give us more freedom to evolve the new private function as desired without having to worry about accidentally introducing a subtle breaking change in the logic, if not the return type. And the `find_by_media_type()` would be frozen going forward, so there really isn't much of an additional maintenance burden to speak of.

That being said, we may not do much with `resolve()` going forward, either. Which would remove the pros from my argument. 😄

Reply:
I guess it would not hurt leaving it as is, since it's on the chopping blocks for v4